File: kakao.app/js/ked/view/cells.js
var _k_ = {list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}}

var cells

import theme from "../theme.js"

import util from "../util/util.js"
import color from "../util/color.js"


cells = (function ()
{
    function cells (screen)
    {
        this.screen = screen
    
        this["draw_frame"] = this["draw_frame"].bind(this)
        this["fill_rect"] = this["fill_rect"].bind(this)
        this["bg_fill"] = this["bg_fill"].bind(this)
        this["bg_rect"] = this["bg_rect"].bind(this)
        this["posForEvent"] = this["posForEvent"].bind(this)
        this["screenForPos"] = this["screenForPos"].bind(this)
        this["posForScreen"] = this["posForScreen"].bind(this)
        this["isOutsideEvent"] = this["isOutsideEvent"].bind(this)
        this["isInsideEvent"] = this["isInsideEvent"].bind(this)
        this["isOutsideScreen"] = this["isOutsideScreen"].bind(this)
        this["isInsideScreen"] = this["isInsideScreen"].bind(this)
        this["isOutsidePos"] = this["isOutsidePos"].bind(this)
        this["isInsidePos"] = this["isInsidePos"].bind(this)
        this["get_char"] = this["get_char"].bind(this)
        this["set_fg"] = this["set_fg"].bind(this)
        this["set_bg"] = this["set_bg"].bind(this)
        this["set_char"] = this["set_char"].bind(this)
        this["set"] = this["set"].bind(this)
        this["add"] = this["add"].bind(this)
        this["layout"] = this["layout"].bind(this)
        this.x = this.y = this.cols = this.rows = 0
    }

    cells.prototype["rect"] = function ()
    {
        return [this.x,this.y,this.x + this.cols - 1,this.y + this.rows - 1]
    }

    cells.prototype["layout"] = function (x, y, cols, rows)
    {
        this.x = x
        this.y = y
        this.cols = cols
        this.rows = rows
    }

    cells.prototype["wx"] = function (x)
    {
        return (x < 0 ? this.x + this.cols + x : this.x + x)
    }

    cells.prototype["wy"] = function (y)
    {
        return (y < 0 ? this.y + this.rows + y : this.y + y)
    }

    cells.prototype["inside"] = function (x, y)
    {
        return (0 <= x && x < this.cols) && (0 <= y && y < this.rows)
    }

    cells.prototype["add"] = function (x, y, char, fg, bg)
    {
        if (this.inside(x,y))
        {
            return this.screen.add(this.wx(x),this.wy(y),char,fg,bg)
        }
        else
        {
            return Infinity
        }
    }

    cells.prototype["set"] = function (x, y, char, fg, bg)
    {
        if (this.inside(x,y))
        {
            return this.screen.set(this.wx(x),this.wy(y),char,fg,bg)
        }
    }

    cells.prototype["set_char"] = function (x, y, char)
    {
        if (this.inside(x,y))
        {
            return this.screen.set_char(this.wx(x),this.wy(y),char)
        }
    }

    cells.prototype["set_bg"] = function (x, y, bg)
    {
        if (this.inside(x,y))
        {
            return this.screen.set_bg(this.wx(x),this.wy(y),bg)
        }
    }

    cells.prototype["set_fg"] = function (x, y, fg)
    {
        if (this.inside(x,y))
        {
            return this.screen.set_fg(this.wx(x),this.wy(y),fg)
        }
    }

    cells.prototype["get_char"] = function (x, y)
    {
        return this.screen.get_char(this.wx(x),this.wy(y))
    }

    cells.prototype["isInsidePos"] = function (x, y)
    {
        var _a_ = util.pos(x,y); x = _a_[0]; y = _a_[1]

        return ((0 <= x && x < this.cols)) && ((0 <= y && y < this.rows))
    }

    cells.prototype["isOutsidePos"] = function (x, y)
    {
        var _a_ = util.pos(x,y); x = _a_[0]; y = _a_[1]

        return x < 0 || x >= this.cols || y < 0 || y >= this.rows
    }

    cells.prototype["isInsideScreen"] = function (x, y)
    {
        return this.isInsidePos(this.posForScreen(x,y))
    }

    cells.prototype["isOutsideScreen"] = function (x, y)
    {
        return this.isOutsidePos(this.posForScreen(x,y))
    }

    cells.prototype["isInsideEvent"] = function (evt)
    {
        return this.isInsidePos(this.posForEvent(evt))
    }

    cells.prototype["isOutsideEvent"] = function (evt)
    {
        return this.isOutsidePos(this.posForEvent(evt))
    }

    cells.prototype["posForScreen"] = function (x, y)
    {
        var _a_ = util.pos(x,y); x = _a_[0]; y = _a_[1]

        return [x - this.x,y - this.y]
    }

    cells.prototype["screenForPos"] = function (x, y)
    {
        var _a_ = util.pos(x,y); x = _a_[0]; y = _a_[1]

        return [x + this.x,y + this.y]
    }

    cells.prototype["posForEvent"] = function (evt)
    {
        return this.posForScreen(evt.cell)
    }

    cells.prototype["bg_rect"] = function (x1, y1, x2, y2, bg)
    {
        var col, row

        if (x1 < 0)
        {
            x1 = this.cols + x1
        }
        if (x2 < 0)
        {
            x2 = this.cols + x2
        }
        if (y1 < 0)
        {
            y1 = this.rows + y1
        }
        if (y2 < 0)
        {
            y2 = this.rows + y2
        }
        for (var _a_ = row = y1, _b_ = y2; (_a_ <= _b_ ? row <= y2 : row >= y2); (_a_ <= _b_ ? ++row : --row))
        {
            if (row < this.rows)
            {
                for (var _c_ = col = x1, _d_ = x2; (_c_ <= _d_ ? col <= x2 : col >= x2); (_c_ <= _d_ ? ++col : --col))
                {
                    if (col < this.cols)
                    {
                        this.set_bg(col,row,bg)
                    }
                }
            }
        }
    }

    cells.prototype["bg_fill"] = function (x1, y1, x2, y2, bg)
    {
        var col, row

        if (x1 < 0)
        {
            x1 = this.cols + x1
        }
        if (x2 < 0)
        {
            x2 = this.cols + x2
        }
        if (y1 < 0)
        {
            y1 = this.rows + y1
        }
        if (y2 < 0)
        {
            y2 = this.rows + y2
        }
        for (var _a_ = row = y1, _b_ = y2; (_a_ <= _b_ ? row <= y2 : row >= y2); (_a_ <= _b_ ? ++row : --row))
        {
            if (row < this.rows)
            {
                for (var _c_ = col = x1, _d_ = x2; (_c_ <= _d_ ? col <= x2 : col >= x2); (_c_ <= _d_ ? ++col : --col))
                {
                    if (col < this.cols)
                    {
                        this.set_bg(col,row,bg)
                        this.set_char(col,row,' ')
                    }
                }
            }
        }
    }

    cells.prototype["fill_rect"] = function (x1, y1, x2, y2, char, fg, bg)
    {
        var col, row

        if (x1 < 0)
        {
            x1 = this.cols + x1
        }
        if (x2 < 0)
        {
            x2 = this.cols + x2
        }
        if (y1 < 0)
        {
            y1 = this.rows + y1
        }
        if (y2 < 0)
        {
            y2 = this.rows + y2
        }
        for (var _a_ = row = y1, _b_ = y2; (_a_ <= _b_ ? row <= y2 : row >= y2); (_a_ <= _b_ ? ++row : --row))
        {
            for (var _c_ = col = x1, _d_ = x2; (_c_ <= _d_ ? col <= x2 : col >= x2); (_c_ <= _d_ ? ++col : --col))
            {
                this.set(col,row,char,fg,bg)
            }
        }
    }

    cells.prototype["draw_frame"] = function (x1, y1, x2, y2, opt)
    {
        var bg, fg, x, y, _122_16_, _130_20_, _131_20_

        if (x1 < 0)
        {
            x1 = this.cols + x1
        }
        if (x2 < 0)
        {
            x2 = this.cols + x2
        }
        if (y1 < 0)
        {
            y1 = this.rows + y1
        }
        if (y2 < 0)
        {
            y2 = this.rows + y2
        }
        opt = (opt != null ? opt : {})
        opt.pad = ((_122_16_=opt.pad) != null ? _122_16_ : [1,0])
        fg = ((_130_20_=opt.fg) != null ? _130_20_ : '#888888')
        bg = ((_131_20_=opt.bg) != null ? _131_20_ : null)
        this.set(x1,y1,'╭',fg,bg)
        this.set(x2,y1,'╮',fg,bg)
        this.set(x1,y2,'╰',fg,bg)
        this.set(x2,y2,'╯',fg,bg)
        this.fill_rect(x1 + 1,y1,x2 - 1,y1,'─',fg,bg)
        this.fill_rect(x1 + 1,y2,x2 - 1,y2,'─',fg,bg)
        this.fill_rect(x1,y1 + 1,x1,y2 - 1,'│',fg,bg)
        this.fill_rect(x2,y1 + 1,x2,y2 - 1,'│',fg,bg)
        for (var _a_ = x = 0, _b_ = opt.pad[0]; (_a_ <= _b_ ? x < opt.pad[0] : x > opt.pad[0]); (_a_ <= _b_ ? ++x : --x))
        {
            this.fill_rect(x1 + 1 + x,y1 + 1,x1 + 1 + x,y2 - 1,' ',fg,bg)
            this.fill_rect(x2 - 1 - x,y1 + 1,x2 - 1 - x,y2 - 1,' ',fg,bg)
        }
        var list = _k_.list(opt.hdiv)
        for (var _c_ = 0; _c_ < list.length; _c_++)
        {
            y = list[_c_]
            this.set(x1,y,'├',fg,bg)
            this.set(x2,y,'┤',fg,bg)
            this.fill_rect(x1 + 1,y,x2 - 1,y,'─',fg,bg)
        }
    }

    cells.prototype["draw_path"] = function (x, mx, y, pth, bg)
    {
        var ci, fg, lastDot, lastSlash, si

        lastSlash = pth.lastIndexOf('/')
        lastDot = pth.lastIndexOf('.')
        si = _k_.max(0,pth.length - mx + x)
        for (var _a_ = ci = si, _b_ = pth.length; (_a_ <= _b_ ? ci < pth.length : ci > pth.length); (_a_ <= _b_ ? ++ci : --ci))
        {
            fg = (ci > lastSlash ? 'file' : 'dir')
            fg = theme[fg]
            if ((lastSlash <= lastDot && lastDot <= ci))
            {
                fg = color.darken(fg)
            }
            if (_k_.in(pth[ci],'./'))
            {
                fg = color.darken(fg)
            }
            this.set(x,y,pth[ci],fg,bg)
            x += 1
        }
        return pth.length - si
    }

    return cells
})()

export default cells;

File: kakao.app/love/lua/view/base/cells.lua
--[[
     ███████  ████████  ███      ███       ███████  
    ███       ███       ███      ███      ███       
    ███       ███████   ███      ███      ███████   
    ███       ███       ███      ███           ███  
     ███████  ████████  ███████  ███████  ███████   

    represents a rectangular set of character cells
    provides some utilities to quickly draw cells
--]]

belt = require "edit.tool.belt"
color = require "theme.color"
theme = require "theme.theme"


local cells = class("cells")
    


function cells:init() 
        self.x = 1
        self.y = 1
        self.cols = 0
        self.rows = 0
        return self
    end


function cells:__tostring() 
    return "[cells " .. tostring(self.x) .. " " .. tostring(self.y) .. " " .. tostring(self.cols) .. " " .. tostring(self.rows) .. "]"
    end


function cells:rect() 
    return array(self.x, self.y, ((self.x + self.cols) - 1), ((self.y + self.rows) - 1))
    end


function cells:layout(x, y, cols, rows) 
        self.x = x
        self.y = y
        self.cols = cols
        self.rows = rows
        self.c = belt.cells(self.cols, self.rows)
        return self.c
    end


function cells:size() 
    return array(self.cols, self.rows)
    end

--  0000000  00000000  000000000  
-- 000       000          000     
-- 0000000   0000000      000     
--      000  000          000     
-- 0000000   00000000     000     


function cells:wx(x) 
    if (x < 0) then 
    return ((self.x + self.cols) + x) else 
    return (self.x + x)
             end
    end

function cells:wy(y) 
    if (y < 0) then 
    return ((self.y + self.rows) + y) else 
    return (self.y + y)
             end
    end


function cells:inside(x, y) 
    return ((((1 <= x) and (x <= self.cols)) and (1 <= y)) and (y <= self.rows))
    end

function cells:outside(x, y) 
    return not self:inside(x, y)
    end


function cells:add(x, y, char, fg, bg) 
        if self:outside(x, y) then 
            print(":OUTSIDE", x, y)
            return math.huge
        end
        
        local w = 1
        if (w > 1) then 
            if (#char > 4) then 
                char = (('\x1b]66;w=2;' + char) + '\x07')
            end
            
            self:set(x, y, char, fg, bg)
            self:set((x + 1), y, nil, fg, bg)
            return 2
        else 
            self:set(x, y, char, fg, bg)
            return 1
        end
    end


function cells:set(x, y, char, fg, bg) 
        if self:inside(x, y) then 
            self.c[y][x].char = char
            self.c[y][x].fg = (fg or array())
            self.c[y][x].bg = (bg or array())
        end
        
        return self
    end


function cells:set_char(x, y, char) 
        if self:inside(x, y) then 
            self.c[y][x].char = char
        end
        
        return self
    end


function cells:set_ch_fg(x, y, char, fg) 
        if self:inside(x, y) then 
            self.c[y][x].char = char
            self.c[y][x].fg = fg
        end
        
        return self
    end


function cells:set_bg(x, y, bg) 
        if self:inside(x, y) then 
            self.c[y][x].bg = bg
        end
        
        return self
    end


function cells:set_fg(x, y, fg) 
        if self:inside(x, y) then 
            self.c[y][x].fg = fg
        end
        
        return self
    end


function cells:set_fg_bg(x, y, fg, bg) 
        if self:inside(x, y) then 
            self.c[y][x].fg = fg
            self.c[y][x].bg = bg
        end
        
        return self
    end

--  0000000   00000000  000000000  
-- 000        000          000     
-- 000  0000  0000000      000     
-- 000   000  000          000     
--  0000000   00000000     000     


function cells:get_char(x, y) 
        if self:inside(x, y) then 
            return self.c[y][x].char
        else 
            print(":OUTSIDE")
        end
        
        return ''
    end


function cells:get_fg(x, y) 
        if self:inside(x, y) then 
            return self.c[y][x].fg
        end
        
        return array()
    end


function cells:get_bg(x, y) 
        if self:inside(x, y) then 
            return self.c[y][x].bg
        end
        
        return array()
    end

-- 00000000    0000000    0000000  
-- 000   000  000   000  000       
-- 00000000   000   000  0000000   
-- 000        000   000       000  
-- 000         0000000   0000000   


function cells:isInsidePos(x, y) 
                     local x, y = belt.pos(x, y) ; return (((1 <= x) and (x <= self.cols)) and ((1 <= y) and (y <= self.rows)))
    end

function cells:isOutsidePos(x, y) 
                     local x, y = belt.pos(x, y) ; return ((((x < 1) or (x > self.cols)) or (y < 1)) or (y > self.rows))
    end


function cells:isInsideScreen(x, y) 
    return self:isInsidePos(self:posForScreen(x, y))
    end

function cells:isOutsideScreen(x, y) 
    return self:isOutsidePos(self:posForScreen(x, y))
    end


function cells:isInsideEvent(evt) 
    return self:isInsidePos(self:posForEvent(evt))
    end

function cells:isOutsideEvent(evt) 
    return self:isOutsidePos(self:posForEvent(evt))
    end


function cells:posForScreen(x, y) 
                     local x, y = belt.pos(x, y) ; return array(((x - self.x) + 1), ((y - self.y) + 1))
    end

function cells:screenForPos(x, y) 
                     local x, y = belt.pos(x, y) ; return array(((x + self.x) - 1), ((y + self.y) - 1))
    end

function cells:posForEvent(evt) 
        return self:posForScreen(ceil((evt.x / _G.screen.cw)), ceil((evt.y / _G.screen.ch)))
    end

-- 0000000     0000000           00000000   00000000   0000000  000000000  
-- 000   000  000                000   000  000       000          000     
-- 0000000    000  0000          0000000    0000000   000          000     
-- 000   000  000   000          000   000  000       000          000     
-- 0000000     0000000   000000  000   000  00000000   0000000     000     


function cells:bg_rect(x1, y1, x2, y2, bg) 
        if (x1 < 0) then x1 = (self.cols + x1) end
        if (x2 < 0) then x2 = (self.cols + x2) end
        
        if (y1 < 0) then y1 = (self.rows + y1) end
        if (y2 < 0) then y2 = (self.rows + y2) end
        
        for row in iter(y1, y2) do 
            if (row < self.rows) then 
                for col in iter(x1, x2) do 
                    if (col < self.cols) then 
                        self:set_bg(col, row, bg)
                    end
                end
            end
        end
    end

-- ████████  ███  ███      ███    
-- ███       ███  ███      ███    
-- ██████    ███  ███      ███    
-- ███       ███  ███      ███    
-- ███       ███  ███████  ███████


function cells:bg_fill(x1, y1, x2, y2, bg) 
        if (x1 < 0) then x1 = (self.cols + x1) end
        if (x2 < 0) then x2 = (self.cols + x2) end
        
        if (y1 < 0) then y1 = (self.rows + y1) end
        if (y2 < 0) then y2 = (self.rows + y2) end
        
        if (x1 > x2) then return end
        
        for row in iter(y1, y2) do 
            if (row <= self.rows) then 
                for col in iter(x1, x2) do 
                    if (col <= self.cols) then 
                        self:set_bg(col, row, bg)
                        self:set_char(col, row, ' ')
                    end
                end
            end
        end
    end


function cells:fill_rect(x1, y1, x2, y2, char, fg, bg) 
        print("FILLRECT")
        x1 = clamp(0, (self.cols - 1), x1)
        y1 = clamp(0, (self.rows - 1), y1)
        if (x2 < 0) then x2 = (self.cols + x2) end
        if (y2 < 0) then y2 = (self.rows + y2) end
        
        for row in iter(y1, y2) do 
            for col in iter(x1, x2) do 
                self:set(col, row, char, fg, bg)
            end
        end
    end


function cells:fill_row(row, x1, x2, char, fg, bg) 
        print("FILLROW")
        if ((x1 < 1) and (x2 < 1)) then return end
        
        x1 = clamp(1, self.cols, x1)
        x2 = clamp(1, self.cols, x2)
        
        if (x2 < x1) then return end
        
        for col in iter(x1, x2) do 
            self:set(col, row, char, fg, bg)
        end
    end


function cells:fill_col(col, y1, y2, char, fg, bg) 
        if ((y1 < 1) and (y2 < 1)) then return end
        
        y1 = clamp(1, self.rows, y1)
        y2 = clamp(1, self.rows, y2)
        
        if (y2 < y1) then return end
        
        for row in iter(y1, y2) do 
            self:set(col, row, char, fg, bg)
        end
    end

-- ████████  ████████    ███████   ██     ██  ████████
-- ███       ███   ███  ███   ███  ███   ███  ███     
-- ██████    ███████    █████████  █████████  ███████ 
-- ███       ███   ███  ███   ███  ███ █ ███  ███     
-- ███       ███   ███  ███   ███  ███   ███  ████████


function cells:draw_frame(x1, y1, x2, y2, opt) 
        if ((x1 < 1) and (x2 < 1)) then return end
        if ((y1 < 1) and (y2 < 1)) then return end
        
        if (x2 < 0) then x2 = ((self.cols + 1) + x2) end
        if (y2 < 0) then y2 = ((self.rows + 1) + y2) end
        
        opt = opt or ({})
        opt.pad = opt.pad or (array(1, 0)) -- padding at left and right edge by default
        
        -- ╭─┬─╮
        -- │ │ │
        -- ├─┼─┤ 
        -- │ │ │
        -- ╰─┴─╯
        
        local fg = (opt.fg or array(100, 100, 100))
        local bg = (opt.bg or nil)
        
        self:set(x1, y1, '╭', fg, bg)
        self:set(x2, y1, '╮', fg, bg)
        self:set(x1, y2, '╰', fg, bg)
        self:set(x2, y2, '╯', fg, bg)
        
        self:fill_row(y1, (x1 + 1), (x2 - 1), '─', fg, bg)
        self:fill_row(y2, (x1 + 1), (x2 - 1), '─', fg, bg)
        self:fill_col(x1, (y1 + 1), (y2 - 1), '│', fg, bg)
        self:fill_col(x2, (y1 + 1), (y2 - 1), '│', fg, bg)
        
        for x = 1, (opt.pad[1] + 1)-1 do 
            self:fill_col(((x1 + 1) + x), (y1 + 1), (y2 - 1), ' ', fg, bg)
            self:fill_col(((x2 - 1) - x), (y1 + 1), (y2 - 1), ' ', fg, bg)
        end
        
        -- for y in opt.hdiv
        --     @set  x1  y '' fg bg
        --     @set  x2  y '' fg bg
        return --     @fill_row y x1+1 x2-1 '─' fg bg
    end

-- ████████    ███████   ███   ███  ███   ███  ███████    ████████  ███████  
-- ███   ███  ███   ███  ███   ███  ████  ███  ███   ███  ███       ███   ███
-- ███████    ███   ███  ███   ███  ███ █ ███  ███   ███  ███████   ███   ███
-- ███   ███  ███   ███  ███   ███  ███  ████  ███   ███  ███       ███   ███
-- ███   ███   ███████    ███████   ███   ███  ███████    ████████  ███████  


function cells:draw_rounded_border(x1, y1, x2, y2, opt) 
        -- ⮐  if x1 < 0 and x2 < 0
        -- ⮐  if y1 < 0 and y2 < 0
        -- 
        -- if x2 < 0 ➜ x2 = @cols + x2 
        -- if y2 < 0 ➜ y2 = @rows + y2 
        -- 
        -- opt ?= {}
        -- fg     = opt.fg or '#888'
        -- zLayer = opt.zLayer or 1000
        
        self:draw_frame(x1, y1, x2, y2, opt)
        
        -- @img x1 y1 'rounded.border.tl' fg zLayer  
        -- @img x2 y1 'rounded.border.tr' fg zLayer  
        -- @img x1 y2 'rounded.border.bl' fg zLayer  
        -- @img x2 y2 'rounded.border.br' fg zLayer  
        -- 
        -- @img x1+1 y1 'rounded.border.t'  fg zLayer x2-1 
        -- if y1+1 < y2 ➜ @img x1 y1+1 'rounded.border.l'  fg zLayer x1 y2-1  
        -- if x1+1 < x2 ➜ @img x1+1 y2 'rounded.border.lb' fg zLayer          
        -- @img x1+2 y2 'rounded.border.b'  fg zLayer x2-1
        -- if y1+1 < y2 ➜ @img x2 y1+1 'rounded.border.rt' fg zLayer x2       
        return -- if y1+2 < y2 ➜ @img x2 y1+2 'rounded.border.r'  fg zLayer x2 y2-1  
    end


function cells:draw_vertical_padding(x, y, fg, bg, zLayer) 
        zLayer = zLayer or 1000
        
        return -- rounded.place @wx(x) @wy(y) 'rounded.vertical' fg nil nil zLayer bg
    end


function cells:draw_horizontal_padding(x, y, fg, bg, zLayer) 
        zLayer = zLayer or 1000
        
        return -- rounded.place @wx(x) @wy(y) 'rounded.horizontal' fg nil nil zLayer bg
    end

--  ███████  ███   ███  ████████    ███████   ███████   ████████ 
-- ███       ███   ███  ███   ███  ███       ███   ███  ███   ███
-- ███       ███   ███  ███████    ███████   ███   ███  ███████  
-- ███       ███   ███  ███   ███       ███  ███   ███  ███   ███
--  ███████   ███████   ███   ███  ███████    ███████   ███   ███


function cells:draw_rounded_cursor(x, y, fg) 
        local cw = _G.screen.cw
        local ch = _G.screen.ch
        
        local xo = ((self.x - 1) * cw)
        local yo = ((self.y - 1) * ch)
        
        local line_x = (xo + ((x - 1) * cw))
        local line_y = ((yo + ((y - 1) * ch)) + (ch / 2))
        local lr = round((cw / 6))
        local lw = (lr * 2)
        
        love.graphics.setColor(fg[1], fg[2], fg[3])
        
        love.graphics.rectangle("fill", (line_x - lr), ((line_y - (ch / 2)) + lr), lw, (ch - lw))
        love.graphics.circle("fill", line_x, ((line_y - (ch / 2)) + lr), lr)
        return love.graphics.circle("fill", line_x, ((line_y + (ch / 2)) - lr), lr)
    end


function cells:draw_rounded_multi_cursor(x, y, fg) 
        local cw = _G.screen.cw
        local ch = _G.screen.ch
        
        local xo = ((self.x - 1) * cw)
        local yo = ((self.y - 1) * ch)
        
        local line_x = (xo + ((x - 1) * cw))
        local line_y = ((yo + ((y - 1) * ch)) + (ch / 2))
        local lr = round((cw / 10))
        local lw = (lr * 2)
        
        love.graphics.setColor(fg[1], fg[2], fg[3])
        
        love.graphics.rectangle("fill", (line_x - lr), (line_y - (ch / 2)), lw, ch)
        -- love.graphics.circle    "fill" line_x line_y-ch/2 lr
        return -- love.graphics.circle    "fill" line_x line_y+ch/2 lr
    end


function cells:img(x, y, name, fg, zLayer, xe, ye) 
        zLayer = zLayer or 1000
        
        if self:outside(x, y) then return end
        
        return -- rounded.place @wx(x) @wy(y) name fg @wx(xe) @wy(ye) zLayer
    end

--  0000000   0000000          000         0000000   0000000   000   000  000000000  00000000    0000000    0000000  000000000  
-- 000   000  000   000        000        000       000   000  0000  000     000     000   000  000   000  000          000     
-- 000000000  000   000        000        000       000   000  000 0 000     000     0000000    000000000  0000000      000     
-- 000   000  000   000  000   000        000       000   000  000  0000     000     000   000  000   000       000     000     
-- 000   000  0000000     0000000          0000000   0000000   000   000     000     000   000  000   000  0000000      000     


function cells:adjustContrastForHighlight(x, y, highlightColor) 
        local ofg = self:get_fg(x, y)
        if (valid(ofg) and valid(highlightColor)) then 
            return self:set_fg(x, y, color.adjustForBackground(ofg, highlightColor))
        end
    end

-- ████████   ████████  ███   ███  ███████    ████████  ████████ 
-- ███   ███  ███       ████  ███  ███   ███  ███       ███   ███
-- ███████    ███████   ███ █ ███  ███   ███  ███████   ███████  
-- ███   ███  ███       ███  ████  ███   ███  ███       ███   ███
-- ███   ███  ████████  ███   ███  ███████    ████████  ███   ███


function cells:render() 
        local lg = love.graphics
        
        local cw = _G.screen.cw
        local ch = _G.screen.ch
        
        local xo = ((self.x - 1) * cw)
        local yo = ((self.y - 1) * ch)
        
        for y in iter(1, self.rows) do 
            -- log "LINE #{y}" @c[y]∙len()
            for x in iter(1, self.cols) do 
                local char = self.c[y][x].char
                
                if (self.c[y][x].bg and (#self.c[y][x].bg > 0)) then 
                    lg.setColor((self.c[y][x].bg[1] / 255), (self.c[y][x].bg[2] / 255), (self.c[y][x].bg[3] / 255))
                    lg.rectangle("fill", (xo + ((x - 1) * cw)), (yo + ((y - 1) * ch)), cw, ch)
                end
                
                if ((#self.c[y][x].fg > 0) and is(self.c[y][x].fg, array)) then 
                    lg.setColor((self.c[y][x].fg[1] / 255), (self.c[y][x].fg[2] / 255), (self.c[y][x].fg[3] / 255))
                end
                
                lg.print(char, (xo + ((x - 1) * cw)), (yo + ((y - 1) * ch)))
            end
        end
        
        return self
    end

return cells